### PR TITLE
Implement user impersonation

### PR DESCRIPTION
**Pull request by SomtochiAma**
Proof of concept for https://github.com/fluxcd/flux2/pull/582
TLDR: This pull request enables the controller to [impersonate](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation) a different user when applying and validating manifests from Sources. This reduces the privileges needed by the controller serviceaccount and also moves the ball towards multi-tenancy.
 
###  Changes
- User impersonation in the controller is implemented behind the `--user-impersonation` flag.
- The `Principal` field is added to the the API as an alternative to `ServiceAccountName`.
```
type Principal struct {
	// Kind specifies the kind of object to be impersonated
	// The kind could be 'User' or 'ServiceAccount'
	// +kubebuilder:validation:Enum=ServiceAccount;User
	Kind string `json:"kind"`

	// The name of the object to be impersonated
	Name string `json:"name"`
}
```
The kind can be either `User` or `ServiceAccount`.
When Principal is not set and the `--user-impersonation` flag is set, a default user `reconciler` is used.

Token impersonation will still be used if:
- when the `user-impersonation` flag is not set
- when `serviceAccountName` is set and `principal` is unset regardless of flag passed in by users

When both `serviceAccountName` and `principal` is set, the flag set by the user determines the type of impersonation. If `--user-impersonation` isn't set, token impersonation is user

For remote clusters, user impersonation is only enabled when the `user-impersonation`  flag is set and `principal` is also set.

## Example
The  user `flux:user:{{namespace}}:{{spec.user}}` in this case `flux:user:tenant-a:admin`  has `cluster-admin` permissions only within the `tenant-a` namespace and the kustomize-controller can only apply manifest in the `tenant-a` namespace when impersonating this user.

```YAML
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: admin
  namespace: tenant-a
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - kind: User
    name: flux:user:tenant-a:admin
    namespace: tenant-a
---
apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
metadata:
  name: podinfo
  namespace: tenant-a
spec:
  targetNamespace: tenant-a
  interval: 5m0s
  path: ./kustomize
  prune: true
  sourceRef:
    kind: GitRepository
    name: podinfo
    namespace: flux-system
  validation: client
  principal:
    kind: User
    name: admin
```

## Feedback & Testing
1. Install flux
```
flux install
```
2. Apply the CRDs from this repository
```
kubectl apply -k https://github.com/SomtochiAma/kustomize-controller/config/crd?ref=user-multi-tenancy
```
3. Create ClusterRole to allow the kustomize-controller to impersonate users and service accounts.
```
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: ServiceAccount
metadata:
  name: user-impersonation
  namespace: flux-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: null
  name: user-impersonation-role
rules:
  - apiGroups:
    - ""
    resources:
      - configmaps
    verbs:
      - get
      - list
      - watch
      - create
      - update
      - patch
      - delete
  - apiGroups:
      - ""
    resources:
      - configmaps/status
    verbs:
      - get
      - update
      - patch
  - apiGroups:
      - ""
    resources:
      - events
    verbs:
      - create
  - apiGroups:
      - "coordination.k8s.io"
    resources:
      - leases
    verbs:
      - get
      - list
      - watch
      - create
      - update
      - patch
      - delete
  - apiGroups:
      - ""
    resources:
      - events
    verbs:
      - create
      - patch
  - apiGroups:
      - ""
    resources:
      - secrets
      - serviceaccounts
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - kustomize.toolkit.fluxcd.io
    resources:
      - kustomizations
    verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
  - apiGroups:
      - kustomize.toolkit.fluxcd.io
    resources:
      - kustomizations/finalizers
    verbs:
      - create
      - delete
      - get
      - patch
      - update
  - apiGroups:
      - kustomize.toolkit.fluxcd.io
    resources:
      - kustomizations/status
    verbs:
      - get
      - patch
      - update
  - apiGroups:
      - source.toolkit.fluxcd.io
    resources:
      - buckets
      - gitrepositories
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - source.toolkit.fluxcd.io
    resources:
      - buckets/status
      - gitrepositories/status
    verbs:
      - get
  - apiGroups:
    - ""
    resources:
      - groups
      - users
      - serviceaccounts
    verbs:
      - impersonate
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: user-impersonation
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: user-impersonation-role
subjects:
  - kind: ServiceAccount
    name: user-impersonation
    namespace: flux-system
EOF
```

4.  Patch the kustomize-controller deployment so that it uses an image build of this PR,  enables user-impersonation and runs under the more restricted service account
```
kubectl patch deploy/kustomize-controller -n flux-system -p "$(cat << EOF
spec:
  template:
    spec:
      containers:
        - name: manager
          args:
          - --watch-all-namespaces=true
          - --log-level=info
          - --log-encoding=json
          - --enable-leader-election
          - --user-impersonation
          image: somma/kustomize-controller:user-imp-4c6938415
      serviceAccountName: user-impersonation
EOF
)"
```
Signed-off-by: Somtochi Onyekwere <somtochionyekwere@gmail.com>

### CODE REVIEW

File: config/crd/bases/kustomize.toolkit.fluxcd.io_kustomizations.yaml [hunk 1]
```diff
@@ -305,10 +323,6 @@ spec:
                   When not specified, the controller uses the KustomizationSpec.Interval
                   value to retry failures.
                 type: string
-              serviceAccountName:
```
Comment (stefanprodan) [inline config/crd/bases/kustomize.toolkit.fluxcd.io_kustomizations.yaml:308]: We can't remove this filed from the `v1beta1` API. If `principal` is `nil` and if `serviceAccountName` is specified, the controller should behave like `principal` is set to that service account name.

File: config/testdata/impersonation/test.yaml [hunk 2]
```diff
@@ -51,7 +51,9 @@ metadata:
   namespace: impersonation
 spec:
   targetNamespace: impersonation
-  serviceAccountName: gotk-reconciler
```
Comment (stefanprodan) [inline config/testdata/impersonation/test.yaml:54]: Undo this change, as I said, we can't break the current implementation.

File: config/testdata/user-impersonation/test.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+apiVersion: v1
```
Comment (stefanprodan) [inline config/testdata/user-impersonation/test.yaml:1]: Rename this file to `use-test.yaml`

File: config/testdata/user-impersonation/test2.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+apiVersion: v1
```
Comment (stefanprodan) [inline config/testdata/user-impersonation/test2.yaml:1]: Rename this file to `sa-test.yaml`

File: .github/workflows/e2e.yaml [hunk 5]
```diff
@@ -80,10 +80,14 @@ jobs:
...
-          kubectl -n tenant-a apply -f ./config/testdata/user-impersonation/test.yaml
+          kubectl -n tenant-a apply -f ./config/testdata/user-impersonation/use-test.yaml
           kubectl -n tenant-a wait kustomizations/podinfo --for=condition=ready --timeout=4m
-          kubectl -n tenant-b apply -f ./config/testdata/user-impersonation/test2.yaml
+          kubectl -n tenant-a  delete kustomizations/podinfo --wait
+          kubectl -n tenant-b apply -f ./config/testdata/user-impersonation/sa-test.yaml
           kubectl -n tenant-b wait kustomizations/podinfo --for=condition=ready --timeout=4m
+          kubectl -n tenant-b  delete kustomizations/podinfo --wait
+          kubectl -n tenant-c apply -f ./config/testdata/user-impersonation/fail-sa-test.yaml
+          until kubectl -n kustomize-system logs deploy/kustomize-controller 2>&1 | grep "Error from server (Forbidden)" ; do sleep 2; done
```
Comment (stefanprodan) [inline .github/workflows/e2e.yaml:90]: Replace the logs search with `kubectl -n tenant-c get kustomization podinfo -oyaml`, as the `Forbidden` error should be in the Kustomization status message.

File: config/testdata/user-impersonation/patch.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+spec:
+  template:
+    spec:
+      containers:
+        - name: manager
+          args:
+          - --events-addr=http://notification-controller/
```
Comment (stefanprodan) [inline config/testdata/user-impersonation/patch.yaml:7]: Please remove all these args and use only those specified here https://github.com/fluxcd/kustomize-controller/blob/main/config/manager/deployment.yaml#L44-L47

File: config/testdata/user-impersonation/token-imp.yaml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    tag: "5.0.3"
+---
+apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
+kind: Kustomization
+metadata:
+  name: podinfo
+  namespace: tenant-d
+spec:
+  targetNamespace: tenant-d
+  # reconciler should use token impersonation when principal is nil ans serviceAccountName is specifiged
```
Comment (stefanprodan) [inline config/testdata/user-impersonation/token-imp.yaml:54]:
```suggestion
  # reconciler should use token impersonation when principal is nil and serviceAccountName is specified
```

File: config/testdata/user-impersonation/token-imp.yaml [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  sourceRef:
+    kind: GitRepository
+    name: podinfo
+    namespace: tenant-d
+  validation: client
+  timeout: 2m
+  healthChecks:
+    - kind: Deployment
+      name: podinfo
+      namespace: tenant-d
```
Comment (stefanprodan) [inline config/testdata/user-impersonation/token-imp.yaml:68]: New line here

File: config/testdata/user-impersonation/patch.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+          - --watch-all-namespaces=true
+          - --log-level=info
+          - --log-encoding=json
+          - --enable-leader-election
+          - --user-impersonation
```
Comment (stefanprodan) [inline config/testdata/user-impersonation/patch.yaml:12]: New line here

File: controllers/kustomization_controller.go [hunk 10]
```diff
@@ -21,6 +21,7 @@ import (
 	"errors"
 	"fmt"
 	"io/ioutil"
+	"k8s.io/client-go/rest"
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:24]: Move this to the section below.

File: controllers/kustomization_controller.go [hunk 11]
```diff
@@ -865,3 +843,77 @@ func (r *KustomizationReconciler) patchStatus(ctx context.Context, req ctrl.Requ
...
+	}
+
+	if kustomization.Spec.Principal == nil {
+		return cmd, nil
+	}
+
+	var saName string
+	if kustomization.Spec.ServiceAccountName != "" {
+		saName = kustomization.Spec.ServiceAccountName
+	} else if kustomization.Spec.Principal.Kind == "ServiceAccount" {
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:899]: Add the `ServiceAccount` and `User` strings to the runtime package as vars and use them in this PR please.

File: main.go [hunk 12]
```diff
@@ -135,6 +138,7 @@ func main() {
 		ExternalEventRecorder: eventRecorder,
 		MetricsRecorder:       metricsRecorder,
 		StatusPoller:          polling.NewStatusPoller(mgr.GetClient(), mgr.GetRESTMapper()),
+		EnableFluxUsers:       enableFluxUsers,
```
Comment (stefanprodan) [inline main.go:141]: Please rename `EnableFluxUsers` to `EnableUserImpersonation`

File: controllers/kustomization_controller.go [hunk 13]
```diff
@@ -865,3 +843,81 @@ func (r *KustomizationReconciler) patchStatus(ctx context.Context, req ctrl.Requ
...
+		kustomization.Spec.Principal = &kustomizev1.Principal{
+			Name: kustomization.Spec.ServiceAccountName,
+			Kind: pkgclient.ServiceAccountType,
+		}
+	}
+
+	if enableFluxUsers {
+		var user string
+		if kustomization.Spec.Principal == nil {
+			user = fmt.Sprintf("flux:user:%s:%s", kustomization.Namespace, "reconciler")
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:873]: Please extract `reconciler` as a var in fluxcd/pkg

File: controllers/kustomization_controller.go [hunk 14]
```diff
@@ -865,3 +843,81 @@ func (r *KustomizationReconciler) patchStatus(ctx context.Context, req ctrl.Requ
 
 	return r.Status().Patch(ctx, &kustomization, patch)
 }
+
+func (r *KustomizationReconciler) buildKubectlCmdImpersonation(ctx context.Context, kustomization kustomizev1.Kustomization, imp *KustomizeImpersonation, cmd string) (string, error) {
+	enableFluxUsers := r.EnableUserImpersonation
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:848]: Please replace `enableFluxUsers` with `enableUserImpersonation` everywhere

File: controllers/kustomization_controller.go [hunk 15]
```diff
@@ -865,3 +843,81 @@ func (r *KustomizationReconciler) patchStatus(ctx context.Context, req ctrl.Requ
...
+		return cmd, nil
+	}
+
+	var saName string
+	if kustomization.Spec.ServiceAccountName != "" {
+		saName = kustomization.Spec.ServiceAccountName
+	} else if kustomization.Spec.Principal.Kind == pkgclient.ServiceAccountType {
+		saName = kustomization.Spec.Principal.Kind
+	} else if kustomization.Spec.Principal.Kind == pkgclient.UserType {
+		return "", errors.New("cannot impersonate user if --enable-flux-users is not set")
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:906]: There is no `--enable-flux-users` flag
